File: parser.c
/* Ponto de entrada do analisador sintático do MINIC em C, como o enunciado
 * pede:
 *
 *     gcc -Wall -Wextra -std=c11 parser.c -o parser
 *     ./parser codigo.c
 *
 * O código de verdade mora em src/c/ (lexer.c da etapa 1, mais ast.c, parser.c
 * e parser_main.c da etapa 2). Este arquivo existe porque o script de teste do
 * professor compila **uma única unidade de tradução**:
 *
 *     gcc -Wall -Wextra -std=c11 $PARSER -o "${PARSER/.c/}"
 *
 * ou seja, ele passa um único .c para o compilador. Incluir os fontes aqui faz
 * com que essa linha de comando funcione sem adaptação nenhuma — o que importa,
 * porque é o professor quem vai rodá-la.
 *
 * Para desenvolver, a compilação separada continua valendo e é preferível (o
 * compilador confere cada arquivo isoladamente):
 *
 *     gcc -Wall -Wextra -std=c11 src/c/lexer.c src/c/ast.c src/c/parser.c \
 *         src/c/parser_main.c -o src/c/parser
 */

#include "src/c/lexer.c"
#include "src/c/ast.c"
#include "src/c/parser.c"
#include "src/c/parser_main.c"

File: parser.py
#!/usr/bin/env python3
"""Ponto de entrada do analisador sintático do MINIC, como o enunciado pede.

    python parser.py codigo.c

O código de verdade mora em `src/python/` (`parser.py` e `minic_ast.py`, que
reaproveitam o `lexer.py` da etapa 1); este arquivo existe só para que o
comando acima funcione a partir da raiz do repositório, que é como a atividade
manda chamar o parser — e é o que o script `testar_parser_python.sh` do
professor espera encontrar.

Aceita os mesmos argumentos do módulo (`--tree`, `--tokens`, `--help`).
"""

import importlib.util
import sys
from pathlib import Path

# Carrega src/python/parser.py pelo caminho do arquivo, e não pelo nome
# `parser`: assim não há como o import pegar este mesmo arquivo, um módulo
# `parser` de outra origem (em Python <= 3.9 havia um embutido) ou depender do
# diretório de onde o comando foi executado.
_ARQUIVO = Path(__file__).resolve().parent / "src" / "python" / "parser.py"
_spec = importlib.util.spec_from_file_location("minic_parser", _ARQUIVO)
_modulo = importlib.util.module_from_spec(_spec)
_spec.loader.exec_module(_modulo)

if __name__ == "__main__":
    sys.exit(_modulo.main(sys.argv))
